### PR TITLE
feat: Added simple polygon area function

### PR DESCRIPTION
**Pull request by waffleSheep**
#### Description of Change

Algorithm that takes a set of points of a simple polygon and returns the area enclosed by that polygon 

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1489"></a>

### CODE REVIEW

File: geometry/polygon_area.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                       std::abs(first_term - second_term + last_point->x * first_point->y - first_point->x * last_point->y);
+            }
+        };
+    }   // namespace polygon_area
+}  // namespace geometry
+
+/**
+ * This is the main function for testing the algorithm
+ */
+auto main() -> int {
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:78]:
```suggestion
int main {
```

File: geometry/polygon_area.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    //Area of polygon enclosed by points (3, 4), (5, 11), (12, 8), (9, 5), (5, 6). Expected value is 30
+    std::vector<geometry::polygon_area::Point<int>> v2 = {{3, 4},
+                                                          {5, 11},
+                                                          {12, 8},
+                                                          {9, 5},
+                                                          {5, 6}};
+    geometry::polygon_area::Polygon_Area<int> p2(v2);
+    assert(p2.get_polygon_area() == 30);
+    std::cout << "Test 3: passed" << std::endl;
+}
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:103]:
```suggestion
}

```

File: geometry/polygon_area.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                return 0.5 *
+                       std::abs(first_term - second_term + last_point->x * first_point->y - first_point->x * last_point->y);
+            }
+        };
+    }   // namespace polygon_area
+}  // namespace geometry
+
+/**
+ * This is the main function for testing the algorithm
+ */
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:79]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

**Comment by waffleSheep** [comment]
I have enabled GitHub actions. Thank you!

**Review by lodenrogue — DISMISSED** [review]

File: geometry/polygon_area.cpp [hunk 4]
```diff
@@ -0,0 +1,110 @@
+/**
+ * @file
+ * @brief Implementation of [Shoelace
+ * formula](https://en.wikipedia.org/wiki/Shoelace_formula).
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:4]: Is this even the implementation of the Shoelace Formula? 😅
```suggestion
 * @brief Implementation of [Polygon Area](https://en.wikipedia.org/wiki/https://en.wikipedia.org/wiki/Polygon).
```

File: geometry/polygon_area.cpp [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+#include <vector>    /// for vector data structure
+
+/**
+ *  @namespace geometry
+ *  @brief Geometry algorithms
+ */
+namespace geometry {
+/**
+ *  @namespace polygon_area
+ *  @brief polygon area functions
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:25]:
```suggestion
 *  @brief Functions for [Polygon Area](https://en.wikipedia.org/wiki/https://en.wikipedia.org/wiki/Polygon) implementation.
```

File: geometry/polygon_area.cpp [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+ */
+namespace polygon_area {
+/**
+ * Point Struct
+ * @tparam for numeric type of the point
+ */
+template <typename T>
+struct Point {
+    T x, y;  /// x and y values for a point
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:34]:
```suggestion
    T x, y;  /// `x` and `y` values for a point
```

File: geometry/polygon_area.cpp [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+};
+
+/**
+ * Class for Polygon Area
+ * @tparam for numeric type of the points
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:39]:
```suggestion
 * @tparam for the numeric type of the points
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: geometry/polygon_area.cpp [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+#include <vector>    /// for vector data structure
+
+/**
+ *  @namespace geometry
+ *  @brief Geometry algorithms
+ */
+namespace geometry {
+/**
+ *  @namespace polygon_area
+ *  @brief Functions for [Polygon
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:25]:
```suggestion
 * @namespace polygon_area
 * @brief Functions for [Polygon
```

File: geometry/polygon_area.cpp [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+ * Area](https://en.wikipedia.org/wiki/https://en.wikipedia.org/wiki/Polygon)
+ * implementation.
+ */
+namespace polygon_area {
+/**
+ * Point Struct
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:31]: Please provide a better brief description.

File: geometry/polygon_area.cpp [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+ * @tparam for numeric type of the point
+ */
+template <typename T>
+struct Point {
+    T x, y;  /// `x` and `y` values for a point
+};
+
+/**
+ * Class for Polygon Area
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:40]: Please provide a better brief description.

File: geometry/polygon_area.cpp [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+ * @tparam for the numeric type of the points
+ */
+template <typename T>
+class Polygon_Area {
+    std::vector<Point<T>> points;  /// list of all points in the 2d space
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:45]:
```suggestion
    std::vector<Point<T>> points;  ///< list of all points in the 2d space
```

File: geometry/polygon_area.cpp [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+ public:
+    /**
+     * Constructor of given class
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:48]:
```suggestion
     * Constructor of a given class
```

File: geometry/polygon_area.cpp [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        /**
+         * This lambda multiplies the x value of the
+         * first point by the y value of the second point
+         */
+        auto multiply_xy_of_two_points = [](const Point<T> &first_point,
+                                            const Point<T> &second_point) {
+            return first_point.x * second_point.y;
+        };
+        auto first_point = points.begin();             // first point iterator
+        auto last_point = std::prev(points.end(), 1);  // last point iterator
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:71]:
```suggestion
        auto first_point = points.begin();             ///< first point iterator
        auto last_point = std::prev(points.end(), 1);  ///< last point iterator
```

File: geometry/polygon_area.cpp [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+        auto first_term =
+            std::inner_product(points.begin(), std::prev(points.end(), 1),
+                               std::next(points.begin(), 1), 0, std::plus<T>(),
+                               multiply_xy_of_two_points);
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:75]: Missing one-line documentation.

File: geometry/polygon_area.cpp [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    // Area of polygon enclosed by points (0,0), (0,1), (1,1), (2,0). Expected
+    // value is 1.5
+    std::vector<geometry::polygon_area::Point<double>> v = {
+        {0.0, 0.0}, {0.0, 1.0}, {1.0, 1.0}, {2.0, 0.0}};
+    geometry::polygon_area::Polygon_Area<double> p(v);
+    assert(p.get_polygon_area() == 1.5);
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:97]: Please create a separate function for tests, and call that function via the `main` function.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Good work.

File: geometry/polygon_area.cpp [hunk 16]
```diff
@@ -0,0 +1,124 @@
+/**
+ * @file
+ * @brief Implementation of [Shoelace
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:3]: Is this the implementation of the Shoelace formula or the Polygon Area? 🤔

File: geometry/polygon_area.cpp [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+        {3, 4}, {5, 11}, {12, 8}, {9, 5}, {5, 6}};
+    geometry::polygon_area::Polygon_Area<int> p2(v2);
+    assert(p2.get_polygon_area() == 30);
+    std::cout << "Test 3: passed" << std::endl;
+}
+
+/**
+ * This is the main function for running the test function
+ */
+int main() { tests(); }
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:124]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
    tests();  // run self-test implementations
    return 0;
}
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: geometry/polygon_area.cpp (hunk 16 above)
Comment (waffleSheep) [reply geometry/polygon_area.cpp:3]: This is a function which uses shoelace formula to calculate the polygon area, much like the jarvis algorithm is used to solve the convex hull problem

File: geometry/polygon_area.cpp [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+                              first_point->x * last_point->y);
+    }
+};
+}  // namespace polygon_area
+}  // namespace geometry
+
+/**
+ * This is the testing function for testing the algorithm
+ */
+void tests() {
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:97]:
```suggestion
static void tests() {
```

File: geometry/polygon_area.cpp [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+                              last_point->x * first_point->y -
+                              first_point->x * last_point->y);
+    }
+};
+}  // namespace polygon_area
+}  // namespace geometry
+
+/**
+ * This is the testing function for testing the algorithm
+ */
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:96]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
```

File: geometry/polygon_area.cpp [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+        auto first_term =
+            std::inner_product(points.begin(), std::prev(points.end(), 1),
+                               std::next(points.begin(), 1), 0, std::plus<T>(),
+                               multiply_xy_of_two_points);
+        // second "shoelace" term, it is the sum of the y value of every point
+        // not including the last point multiplied by the x value of the point
+        // after
+        auto second_term = std::inner_product(
+            std::next(points.begin(), 1), points.end(), points.begin(), 0,
+            std::plus<T>(), multiply_xy_of_two_points);
```
Comment (Panquesito7) [inline geometry/polygon_area.cpp:82]: Please add one-line documentation explaining what does this variable does or is for.

File: geometry/polygon_area.cpp (hunk 14 above)
Comment (Panquesito7) [reply geometry/polygon_area.cpp:76]: Again (already mentioned this above), missing one-line documentation.

File: geometry/polygon_area.cpp (hunk 10 above)
Comment (Panquesito7) [reply geometry/polygon_area.cpp:38]: Still missing a proper description.

File: geometry/polygon_area.cpp (hunk 9 above)
Comment (Panquesito7) [reply geometry/polygon_area.cpp:29]: Still missing a proper description.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! Good work.

**Pull request closed without merging**